src/main/scala/is/hail/expr/Parser.scala

```diff
@@ -650,23 +650,42 @@ object Parser extends JavaTokenParsers {
   
   def referenceGenomeDependentFunction: Parser[String] = "LocusInterval" | "LocusAlleles" | "Locus" | "getReferenceSequence"
 
-  def intervalWithEndpoints[T](bounds: Parser[(T, T)]): Parser[Interval] = {
+  def intervalWithEndpoints[T](bounds: Parser[(T, T, Boolean, Boolean)]): Parser[Interval] = {
     val start = ("[" ^^^ true) | ("(" ^^^ false)
     val end = ("]" ^^^ true) | (")" ^^^ false)
 
-    start ~ bounds ~ end ^^ {case istart ~ int ~ iend => Interval(int._1, int._2, istart, iend)} |
-      bounds ^^ {int => Interval(int._1, int._2, true, false)}
+    start ~ bounds ~ end ^^ { case istart ~ int ~ iend => Interval(int._1, int._2, istart, iend) } |
+      bounds ^^ { int => Interval(int._1, int._2, int._3, int._4) }
   }
 
-  def locusInterval(rg: RGBase): Parser[Interval] = {
+  def locusInterval(rgBase: RGBase): Parser[Interval] = {
+    val rg = rgBase.asInstanceOf[ReferenceGenome]
     val contig = rg.contigParser
-    val valueParser = locus(rg) ~ "-" ~ locus(rg) ^^ { case l1 ~ _ ~ l2 => (l1, l2) } |
-      locus(rg) ~ "-" ~ pos ^^ { case l1 ~ _ ~ p2 => (l1, l1.copyChecked(rg, position = p2.getOrElse(rg.contigLength(l1.contig)))) } |
-      contig ~ "-" ~ contig ^^ { case c1 ~ _ ~ c2 => (Locus(c1, 1, rg), Locus(c2, rg.contigLength(c2), rg)) } |
-      contig ^^ { c => (Locus(c, 1), Locus(c, rg.contigLength(c))) }
-      intervalWithEndpoints(valueParser)
+
+    val valueParser =
+      locusUnchecked(rg) ~ "-" ~ rg.contigParser ~ ":" ~ pos ^^ { case l1 ~ _ ~ c2 ~ _ ~ p2 => p2 match {
```

Comment:
This seems like it has to be too complex, but I'm not exactly sure how to fix it. I guess one option is to always return an interval that includes the start and the end. What do you think of that?

Reply:
I agree it is too complicated, but the reason for that is because we allow the bounds to be specified with "[" and ")" in addition to having a default behavior and allowing the user to specify a proxy for the contig end. 

It's hard to do the right thing in both cases where the bounds are explicitly stated versus not stated. I thought about how to do this better, and couldn't come up with one.

Reply:
ahhh.